app/services/data_migrations/fix_candidate_api_updated_at.rb

```diff
@@ -0,0 +1,28 @@
+module DataMigrations
+  class FixCandidateAPIUpdatedAt
+    TIMESTAMP = 20211208172459
+    MANUAL_RUN = false
+
+    # Ensure that all the candidates have a `candidate_api_updated_at` that is
+    # no earlier than the `created_at` of the most recent application form.
+    def change
+      candidates = Candidate
+        .joins('inner join (
+            select candidates.id id,
+              max(application_forms.created_at) max_form_created_at
+            from candidates
+            inner join application_forms on application_forms.candidate_id = candidates.id
+            group by candidates.id
+          ) candidates_with_latest_application_form on candidates_with_latest_application_form.id = candidates.id'
+        )
+        .select('candidates.*, candidates_with_latest_application_form.max_form_created_at')
+        .where('candidate_api_updated_at < max_form_created_at')
+
+      candidates.find_in_batches(batch_size: 10) do |batch|
+        batch.each do |candidate|
+          candidate.update!(candidate_api_updated_at: candidate.max_form_created_at)
```

Comment:
Is it necessarily correct to update to this? ie is it definitely only missing the form _creation_ event which has gone awry here? Is it possible for the form to be updated in such a way that this timestamp is touched? what about the transition from `unsubmitted_not_started_form` to `unsubmitted_in_progress`?

Reply:
If the query is right we shouldn't be overwriting any values that are later than `max_form_created_at`. All it should be doing is getting the candidates that have a `candidate_api_updated_at < max_form_created_at` and setting them to `max_form_created_at`.

We could go further with this and check for the application form status changes in the audit trail to find out whether there is a later date that `candidate_api_updated_at` should be set to. I'm not sure whether there is a need to do that.